src/value_generators/src/DocumentGenerator.cpp

```diff
@@ -937,6 +938,139 @@ class ObjectIdGenerator : public Generator<bsoncxx::types::b_oid> {
+
+/**
+ * Validate the uuid hex format.
+ * @param hex
+ *   the hex string
+ * @raises InvalidValueGeneratorSyntax if the format is not correct
+ *  The correct format is an 8-4-4-4-12 hex string (case is ignored)
+ */
+void validateUuidHex(std::string hex) {
+
+    auto pos = hex.find_first_not_of(validUuidChars);
+    if ( pos != std::string::npos ) {
+        std::stringstream msg;
+        std::string invalid(hex);
+        invalid.erase(std::remove_if(invalid.begin(), invalid.end(),
+                                 [&](char c) { return validUuidChars.find(c) != std::string::npos; } ),
+                      invalid.end());
+
+        msg << "'" << hex << "' contains invalid characters '" << invalid << "'";
+        BOOST_LOG_TRIVIAL(warning) << " UuidGenerator " << msg.str();
+        BOOST_THROW_EXCEPTION(InvalidValueGeneratorSyntax(msg.str()));
+    }
+
+    std::smatch match;
+    if (!std::regex_search(hex, match, validateUuidRegex)) {
+        std::stringstream msg;
+        msg << "'" << hex << "' is not a valid format. The format must match: '" << validateUuidPattern << "'";
+        BOOST_LOG_TRIVIAL(warning) << " UuidGenerator " << msg.str();
+        BOOST_THROW_EXCEPTION(InvalidValueGeneratorSyntax(msg.str()));
+    }
+}
+
+/**
+ * Validate that the uuid sub type is correct.
+ * @param sub_type
+ *   the binary sub type value
+ * @raises InvalidValueGeneratorSyntax if the sub type is not correct
+*  Currntly only supports sub type 4.
+ */
+void validateUuidSubType(bsoncxx::binary_sub_type sub_type) {
+
+    if (bsoncxx::binary_sub_type::k_uuid != sub_type) {
+        std::stringstream msg;
+        msg << "Invalid binary sub_type. UUID only supports bsoncxx::binary_sub_type::k_uuid("
+            << static_cast<int64_t>(bsoncxx::binary_sub_type::k_uuid) << "), got " << static_cast<int64_t>(sub_type);
+        BOOST_LOG_TRIVIAL(warning) << " UuidGenerator " << msg.str();
+        BOOST_THROW_EXCEPTION(InvalidValueGeneratorSyntax(msg.str()));
+    }
+}
+
+/**
+ * Convert ascii hex nibble to binary.
+ * @param c
+ *   the ascii hex nibble
+ * @return
+ *   the binary version of the hex nibble
+ */
+uint8_t hexNibbleToBin(uint8_t c) {
+    if (c >= '0' && c <= '9') {
+        c -= '0';
+    } else  if (c >= 'a' && c <= 'f') {
+        c = c - 'a' + 10;
+    } else  if (c >= 'A' && c <= 'F') {
+        c = c - 'A' + 10;
+    }
+    return c;
+}
+
+/**
+ * Convert hex string Uuid to binary.
+ * @param hex
+ *   the input hex string, '-' will be ignored.
+ * @param uuid
+ *   the array to store the uuid.
+ * @return the output array
+ */
+uint8_t* hex2BinUuid(std::string hex, uint8_t* uuid) {
+    hex.erase(std::remove_if(hex.begin(), hex.end(),
+                             [&](char c) { return c == '-'; } ),
+              hex.end());
+
+    for(std::string::iterator it = hex.begin(); it != hex.end(); ++it) {
+        int pos = std::distance(hex.begin(), it) / 2;
```

Comment:
Why not just add a counter to the for loop for the index in addition ot the iterator. Then you can increment it once per for loop iteration. It seems like that has to be better than calling std::distances on the string each time.